File: src/apple/Packages/MapleUI/Sources/MapleUI/Molecules/MuiHistogram.swift
// MuiHistogram.swift — Maple UI Molecules-L1 (unified-component-catalog.md
// §2.6; a plot primitive). RGB distribution plot, drawn directly via
// SwiftUI `Canvas`. R/G/B channel colors are literal, not design tokens —
// same "content-functional, not chrome color" precedent as the web
// reference's histogram: a histogram's red/green/blue bars ARE the channel
// identity, independent of the app's accent color. Bars are peak-relative
// (every bin's height is relative to the tallest bin across all three
// channels) with no lane gap — channels overlay each other, unlike Parade's
// side-by-side lanes.

import SwiftUI

public struct MuiHistogram: View {
    /// Pre-binned per-channel counts (e.g. from `MuiHistogramMath.bin`).
    public let r: [Int]
    public let g: [Int]
    public let b: [Int]
    public let width: CGFloat
    public let height: CGFloat

    public init(r: [Int], g: [Int], b: [Int], width: CGFloat = 240, height: CGFloat = 64) {
        self.r = r
        self.g = g
        self.b = b
        self.width = width
        self.height = height
    }

    /// Convenience initializer over raw `0...1` samples, binned internally
    /// via `MuiHistogramMath.bin` — the shape the gallery's fixed sample
    /// data and most real callers (a decoded image's per-channel values)
    /// actually have, rather than pre-binned counts.
    public init(rSamples: [Double], gSamples: [Double], bSamples: [Double], binCount: Int = 64, width: CGFloat = 240, height: CGFloat = 64) {
        self.init(
            r: MuiHistogramMath.bin(rSamples, binCount: binCount),
            g: MuiHistogramMath.bin(gSamples, binCount: binCount),
            b: MuiHistogramMath.bin(bSamples, binCount: binCount),
            width: width,
            height: height
        )
    }

    private static let channelColor = (
        r: Color(red: 220 / 255, green: 80 / 255, blue: 80 / 255).opacity(0.6),
        g: Color(red: 80 / 255, green: 190 / 255, blue: 80 / 255).opacity(0.6),
        b: Color(red: 80 / 255, green: 130 / 255, blue: 220 / 255).opacity(0.6)
    )

    public var body: some View {
        Canvas { context, size in
            let peak = Double([r.max() ?? 1, g.max() ?? 1, b.max() ?? 1].max() ?? 1)
            draw(channel: r, color: Self.channelColor.r, peak: peak, context: context, size: size)
            draw(channel: g, color: Self.channelColor.g, peak: peak, context: context, size: size)
            draw(channel: b, color: Self.channelColor.b, peak: peak, context: context, size: size)
        }
        .frame(width: width, height: height)
        .background(MuiTokens.imageCanvas, in: RoundedRectangle(cornerRadius: MuiTokens.radiusSm, style: .continuous))
        .accessibilityElement(children: .ignore)
        .accessibilityLabel("Histogram")
    }

    private func draw(channel: [Int], color: Color, peak: Double, context: GraphicsContext, size: CGSize) {
        guard !channel.isEmpty, peak > 0 else { return }
        let colWidth = size.width / CGFloat(channel.count)
        for (index, value) in channel.enumerated() {
            let barHeight = CGFloat(Double(value) / peak) * (size.height - 2)
            let rect = CGRect(x: CGFloat(index) * colWidth, y: size.height - barHeight, width: Swift.max(1, colWidth - 0.5), height: barHeight)
            context.fill(Path(rect), with: .color(color))
        }
    }
}

#Preview("MuiHistogram") {
    MuiHistogram(r: (0..<64).map { $0 }, g: (0..<64).map { 64 - $0 }, b: (0..<64).map { i in 32 - abs(32 - i) })
        .padding()
        .background(MuiTokens.bg)
}

File: src/windows/Maple.WinUI/MapleUI/Molecules/MuiHistogram.cs
using System;
using System.Collections.Generic;
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Controls;
using Microsoft.UI.Xaml.Media;
using Microsoft.UI.Xaml.Shapes;
using Windows.UI;

namespace Maple.UI
{
    /// <summary>
    /// Maple.UI Histogram data plot (unified-component-catalog.md §2.6,
    /// "Histogram" row: "RGB distribution plot", built from _(plot
    /// primitive)_) — draws directly via WinUI Shapes onto a
    /// <see cref="Canvas"/> (the same "hand-rolled Shapes plot" technique
    /// <see cref="MuiCurvePlot"/> uses in this library, rather than a Win2D
    /// dependency this project doesn't reference).
    ///
    /// R/G/B channel colors are literal, not theme tokens — same
    /// "content-functional, not chrome color" precedent
    /// `mui-histogram.component.ts`'s own header comment gives (a
    /// histogram's channel colors ARE the channel identity, independent of
    /// the app's accent color). Peak-relative scaling via
    /// <see cref="MuiPlotMath"/>: every bin's bar height is relative to the
    /// tallest bin across all three channels, and channels overlap (not
    /// gapped/laned — that's Parade's layout).
    /// </summary>
    public sealed class MuiHistogram : ContentControl
    {
        public static readonly DependencyProperty RedValuesProperty =
            DependencyProperty.Register(nameof(RedValues), typeof(IReadOnlyList<double>), typeof(MuiHistogram),
                new PropertyMetadata(null, (d, _) => ((MuiHistogram)d).Render()));

        public static readonly DependencyProperty GreenValuesProperty =
            DependencyProperty.Register(nameof(GreenValues), typeof(IReadOnlyList<double>), typeof(MuiHistogram),
                new PropertyMetadata(null, (d, _) => ((MuiHistogram)d).Render()));

        public static readonly DependencyProperty BlueValuesProperty =
            DependencyProperty.Register(nameof(BlueValues), typeof(IReadOnlyList<double>), typeof(MuiHistogram),
                new PropertyMetadata(null, (d, _) => ((MuiHistogram)d).Render()));

        public static readonly DependencyProperty PlotWidthProperty =
            DependencyProperty.Register(nameof(PlotWidth), typeof(double), typeof(MuiHistogram),
                new PropertyMetadata(240.0, (d, _) => ((MuiHistogram)d).Rebuild()));

        public static readonly DependencyProperty PlotHeightProperty =
            DependencyProperty.Register(nameof(PlotHeight), typeof(double), typeof(MuiHistogram),
                new PropertyMetadata(64.0, (d, _) => ((MuiHistogram)d).Rebuild()));

        public IReadOnlyList<double>? RedValues
        {
            get => (IReadOnlyList<double>?)GetValue(RedValuesProperty);
            set => SetValue(RedValuesProperty, value);
        }

        public IReadOnlyList<double>? GreenValues
        {
            get => (IReadOnlyList<double>?)GetValue(GreenValuesProperty);
            set => SetValue(GreenValuesProperty, value);
        }

        public IReadOnlyList<double>? BlueValues
        {
            get => (IReadOnlyList<double>?)GetValue(BlueValuesProperty);
            set => SetValue(BlueValuesProperty, value);
        }

        public double PlotWidth
        {
            get => (double)GetValue(PlotWidthProperty);
            set => SetValue(PlotWidthProperty, value);
        }

        public double PlotHeight
        {
            get => (double)GetValue(PlotHeightProperty);
            set => SetValue(PlotHeightProperty, value);
        }

        private static readonly Color RedColor = Color.FromArgb(0x99, 0xDC, 0x50, 0x50);
        private static readonly Color GreenColor = Color.FromArgb(0x99, 0x50, 0xBE, 0x50);
        private static readonly Color BlueColor = Color.FromArgb(0x99, 0x50, 0x82, 0xDC);

        private readonly Grid _frame = new() { CornerRadius = new CornerRadius(6), BorderThickness = new Thickness(0.5) };
        private readonly Canvas _canvas = new();

        public MuiHistogram()
        {
            _frame.Children.Add(_canvas);
            Content = _frame;
            IsTabStop = false;

            Rebuild();
        }

        private static Brush R(string key) => (Brush)Application.Current.Resources[key];

        private void Rebuild()
        {
            _frame.Width = PlotWidth;
            _frame.Height = PlotHeight;
            _canvas.Width = PlotWidth;
            _canvas.Height = PlotHeight;
            _frame.Background = R("MapleImageCanvas");
            _frame.BorderBrush = R("MapleBorder");
            Render();
        }

        private void Render()
        {
            _canvas.Children.Clear();
            var w = PlotWidth;
            var h = PlotHeight;
            if (w <= 0 || h <= 0) return;

            var channels = new[]
            {
                RedValues ?? Array.Empty<double>(),
                GreenValues ?? Array.Empty<double>(),
                BlueValues ?? Array.Empty<double>(),
            };
            var peak = MuiPlotMath.Peak(channels);

            DrawChannel(RedValues, RedColor, peak, w, h);
            DrawChannel(GreenValues, GreenColor, peak, w, h);
            DrawChannel(BlueValues, BlueColor, peak, w, h);
        }

        private void DrawChannel(IReadOnlyList<double>? values, Color color, double peak, double w, double h)
        {
            var vals = values ?? Array.Empty<double>();
            if (vals.Count == 0) return;

            var barWidth = w / vals.Count;
            var brush = new SolidColorBrush(color);
            for (var i = 0; i < vals.Count; i++)
            {
                var barHeight = MuiPlotMath.BarHeightFraction(vals[i], peak) * h;
                if (barHeight <= 0) continue;

                var rect = new Rectangle { Width = Math.Max(1, barWidth), Height = barHeight, Fill = brush };
                Canvas.SetLeft(rect, i * barWidth);
                Canvas.SetTop(rect, h - barHeight);
                _canvas.Children.Add(rect);
            }
        }
    }
}
